Param defaults - allow many defaults files to be specified

peterbarker opened the pull request:
Closes #5013

khancyr commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -1696,14 +1696,8 @@ bool AP_Param::parse_param_line(char *line, char **vname, float &value)
-  load a default set of parameters from a file
- */
-bool AP_Param::load_defaults_file(const char *filename)
+uint16_t AP_Param::count_defaults_in_file(const char *filename, uint16_t &num_defaults)
```
why change bool in favor of uint16_t ?

peterbarker replied:
Ah.  I was returning the count, but discovered things could fail.  Will patch.

khancyr commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -1776,6 +1761,37 @@ bool AP_Param::load_defaults_file(const char *filename)
+        return false;
+    }
+
+    if (! read_param_defaults_file(filename)) {
```
not important trailling space

peterbarker commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -1789,8 +1799,15 @@ bool AP_Param::load_defaults_file(const char *filename)
+         pname != nullptr;
+         pname = strtok_r(nullptr, ",", &saveptr)) {
+        ::fprintf(stderr, "pname=%s\n", pname);
+        if (! read_param_defaults_file(pname)) {
```
Style thing.  I prefer the space, but will remove.

peterbarker commented:
@khancyr @OXINARF patches pushed.

khancyr commented:
I didn't test but it looks like good!

peterbarker commented:
Per review from @tridge 
 - modified to use strdup rather than allocating string on stack
 - removed a couple of stray fprintfs
 - ran through valgrind (no issues)

lucasdemarchi commented on libraries/AP_Param/AP_Param.cpp:
```diff
@@ -1789,9 +1803,20 @@ bool AP_Param::load_defaults_file(const char *filename)
+        if (!read_param_defaults_file(pname)) {
+            free(mutable_filename);
+            return false;
+        }
```
why are you doing this twice in the same function instead of just reusing the modified string?

peterbarker replied:
Odd, my original response to this question appears to have been lost.

The string is essentially destroyed as part of strtok_r.

lucasdemarchi replied:
what I was thinking was to use some kind of strv loop. We don't have a ready-to-use impl, so it would be more or less this:
https://gist.github.com/lucasdemarchi/1295579f4048f66dbd995864b5c031de

It doesn't allocate twice and has the more or less the same cost of iterating over the string.  Seeing the code now, I have mixed feelings. What do you think?

lucasdemarchi replied:
let's at least compile, humn.
https://gist.github.com/6b1daac9cc9fb41d8b559b5f5689ed77

peterbarker replied:
Well... I actually started writing something along those lines and then thought, "wait, isn't there a library function for just this sort of thing?" :-)

Performance *really*, *really* isn't our concern at this point in the code :-)

I see a lot of pointer manipulation for no huge benefit here.  Perhaps if we did have a strv of some description it would be less intimidating in that respect.

Still, I'm not really fussed either way.  The functionality is good.

peterbarker commented:
On Tue, 15 Nov 2016, Lucas De Marchi wrote:

> +    mutable_filename = strdup(filename);

> why are you doing this twice in the same function instead of just reusing
> the modified string?

I assume you're talking about reinitialising the string after strtok_r?

I don't believe the string is re-usable after you've used it in strtok 
functions.

khancyr commented:
@peterbarker @lucasdemarchi are we good on this ?

peterbarker commented:
@khancyr AFAIK.

lucasdemarchi commented:
Applied. Improvements can be made on top.
